accord-core/src/main/java/accord/local/Command.java

```diff
@@ -19,362 +19,487 @@
-                logger.trace("{}: dep {} is {}", txnId(), command.txnId(), command.status());
-                switch (command.status())
-                {
-                    default:
-                        throw new IllegalStateException();
-                    case NotWitnessed:
-                    case PreAccepted:
-                    case Accepted:
-                    case AcceptedInvalidate:
-                        // we don't know when these dependencies will execute, and cannot execute until we do
-                        command.addListener(this);
-                        addWaitingOnCommit(command);
-                        break;
-                    case Committed:
-                        // TODO: split into ReadyToRead and ReadyToWrite;
-                        //       the distributed read can be performed as soon as those keys are ready, and in parallel with any other reads
-                        //       the client can even ACK immediately after; only the write needs to be postponed until other in-progress reads complete
-                    case ReadyToExecute:
-                    case Executed:
-                    case Applied:
-                        command.addListener(this);
-                        updatePredecessor(command);
-                    case Invalidated:
-                        break;
-                }
-            });
-        }
+        if (!validate(coordinateRanges, executeRanges, shard, route, Check, partialTxn, Add, partialDeps, Set))
+            return CommitOutcome.Insufficient;
 
+        setExecuteAt(executeAt);
+        set(safeStore, coordinateRanges, executeRanges, shard, route, partialTxn, Add, partialDeps, Set);
 
-        boolean isProgressShard = progressKey != null && handles(txnId().epoch, progressKey);
-        commandStore().progressLog().commit(this, isProgressShard, isProgressShard && progressKey.equals(homeKey));
+        setStatus(Committed);
+        logger.trace("{}: committed with executeAt: {}, deps: {}", txnId(), executeAt, partialDeps);
+        populateWaitingOn(safeStore);
 
-        return true;
+        safeStore.progressLog().commit(this, shard);
+
+        // TODO (now): introduce intermediate status to avoid reentry when notifying listeners (which might notify us)
+        maybeExecute(safeStore, shard, true, true);
+        return CommitOutcome.Success;
     }
 
-    public Future<Void> commitAndBeginExecution(Txn txn, Key homeKey, Key progressKey, Timestamp executeAt, Deps deps)
+    protected void populateWaitingOn(SafeCommandStore safeStore)
     {
-        if (!commit(txn, homeKey, progressKey, executeAt, deps))
-            return Writes.SUCCESS;
-
-        return maybeExecute(true);
+        KeyRanges ranges = safeStore.ranges().since(executeAt().epoch);
+        if (ranges != null) {
+            partialDeps().forEachOn(ranges, safeStore.commandStore()::hashIntersects, txnId -> {
+                Command command = safeStore.ifLoaded(txnId);
```

Comment:
Why ifLoaded? If it's not loaded we want to add a listener waiting on commit and rely on the progress log to get it that far?

Reply:
We simply defer processing by registering a listener we immediately ask the command store to invoke?

Reply:
I missed something obvious which is one operates on the store. The other operates on the command.

Reply:
The reason this confuses me is invoking the listener immediately for a command we just loaded is a no-op so why do it?
 
Orthogonally so I can add a comment

Maybe I can summarize to make sure I understand what happens here.

For each dependency if the command isn’t loaded we don’t know the executeAt timestamp so we mark it as waiting on commit  and then add a listener which on any post committed state will update the predecessor state (and not as an insert), and maybe begin execution.

If we already knew about the command then based on its status (pre/post commit) add the listener and then either wait for the commit or update predecessor as an insert.

Reply:
> The reason this confuses me is invoking the listener immediately for a command we just loaded is a no-op so why do it?

Why would it be a no-op? We don't know what its state is, we've just assumed it isn't committed. So we load it to find out what its actual state is, and invoke `onChange` to reflect its actual status in this `Command`

Reply:
OK, I think the reason this was confusing is that the in memory implementation doesn't do that. It's not clear that addAndInvokeListener is loading it async in the Cassandra integration and then invoking the listener.

A comment on what `addAndInvokeListener` does would make it clearer.

Ideally this logic wouldn't be outside the command store you would just say "I want this to happen if the command is loaded or after you load it."

I think this is trying to avoidAddWaitingOnCommit if we already know it's committed?

Reply:
👍

Reply:
This should be resolved now